src/edu/washington/escience/myria/operator/SampledDbInsertTemp.java

```diff
@@ -0,0 +1,221 @@
+  private static final long serialVersionUID = 1L;
+
+  /** The connection to the database database. */
+  private AccessMethod accessMethod;
+  /** The information for the database connection. */
+  private ConnectionInfo connectionInfo;
+  /** The name of the table the tuples should be inserted into. */
+  private final RelationKey sampleRelationKey;
+  /** The name of the table the tuples should be inserted into. */
+  private final RelationKey countRelationKey;
+
+  /** Total number of tuples seen from the child. */
+  private int tupleCount = 0;
+  /** Number of tuples to sample from the stream. */
+  private final int streamSampleSize;
+  /** Reservoir that holds sampleSize number of tuples. */
+  private MutableTupleBuffer reservoir = null;
+  /** Sampled tuples ready to be returned. */
+  private List<TupleBatch> batches;
+  /** Next element of batches List that will be returned. */
+  private int batchNum = 0;
+  /** True if all samples have been gathered from the child. */
+  private boolean doneSamplingFromChild;
+
+  /** The output schema. */
+  private static final Schema COUNT_SCHEMA = Schema.of(
+      ImmutableList.of(Type.INT_TYPE, Type.INT_TYPE, Type.INT_TYPE),
+      ImmutableList.of("WorkerID", "PartitionSize", "StreamSize"));
+
+  /**
+   * @param child
+   *          the source of tuples to be inserted.
+   * @param streamSampleSize
+   *          number of tuples to store from the stream
+   * @param sampleRelationKey
+   *          the key of the table that the tuples should be inserted into.
+   * @param countRelationKey
+   *          the key of the table that the tuple counts will be inserted into.
+   * @param connectionInfo
+   *          the parameters of the database connection.
+   */
+  public SampledDbInsertTemp(final Operator child, final int streamSampleSize,
+      final RelationKey sampleRelationKey, final RelationKey countRelationKey,
+      final ConnectionInfo connectionInfo) {
+    super(child);
+    // Sampling setup.
+    Preconditions.checkArgument(streamSampleSize >= 0L,
+        "sampleSize must be non-negative");
+    this.streamSampleSize = streamSampleSize;
+    doneSamplingFromChild = false;
+
+    // Relation setup.
+    Objects.requireNonNull(sampleRelationKey, "sampleRelationKey");
+    this.sampleRelationKey = sampleRelationKey;
+    Objects.requireNonNull(countRelationKey, "countRelationKey");
+    this.countRelationKey = countRelationKey;
+    this.connectionInfo = connectionInfo;
+  }
+
+  @Override
+  protected TupleBatch fetchNextReady() throws DbException {
+    if (!doneSamplingFromChild) {
+      fillReservoir();
+      batches = reservoir.getAll();
+      // Insert sampled tuples into sampleRelationKey
+      while (batchNum < batches.size()) {
+        TupleBatch batch = batches.get(batchNum);
+        accessMethod.tupleBatchInsert(sampleRelationKey, batch);
+        batchNum++;
+      }
+
+      // Write (WorkerID, PartitionSize, StreamSize) to countRelationKey
```

Comment:
The naming convention comes from the need for SamplingDistribution to know how many tuples were 'pre-sampled' from this stream. I will change the name to PartitionSampleSize to hopefully remove the confusion here.